### PR TITLE
HTML Result: make it a real plugin

### PR DESCRIPTION
**Pull request by clebergnu**
This work rewrites a part of the HTML report generation code, and then moves it out of the `avocado.core`, removing knowledge and dependencies about the HTML code.

Then after that, it makes the HTML result plugin a real (Python-level) plugin, living in a separate module namespace.

### CODE REVIEW

**Comment by clebergnu** [comment]
@ldoktor @apahim review and testing is very much appreciated. Thanks!

File: html/MANIFEST.in [hunk 1]
```diff
@@ -0,0 +1 @@
+recursive-include avocado_result_html/resources *
```
Comment (ldoktor) [inline html/MANIFEST.in]: I don't think we should put this plugin directly as `html`. It's really probable that it'd clash with other `html` module. I think it's fine to put it inside `avocado/plugins` or `avocado-plugins`. Not sure what is a proper place (I know we use `virttest` directly, but this is IMO too much. Consider that this means it'd be directly in `/usr/lib64/python2.6/site-packages`...

Comment (ldoktor) [inline html/MANIFEST.in]: OH, I'm wrong, sorry, I need to take a deeper look...

Comment (clebergnu) [inline html/MANIFEST.in]: I've actually tried to keep it inside `avocado/plugins` but Python/setuptools will try to package it automatically.  Keeping and optional package inside another doesn't seem to be how Python/distutils/setuptools was designed to work.

With regards to the name, it's just one directory name.  We may just as well move it to anything mode complex, say, `results-html-plugin`, but IMHO, that is so easily fixable (just a rename) if a clash *inside* the avocado source tree ever occur.  In the Python "site lib dir", it'll actually use the module name as defined in setuptools: `avocado_result_html`.

Comment (ldoktor) [inline html/MANIFEST.in]: Btw this version does not work for me. I needed to change the paths to:

```
diff --git a/html/avocado_result_html/__init__.py b/html/avocado_result_html/__init__.py
index eaa4a4d..373277d 100644
--- a/html/avocado_result_html/__init__.py
+++ b/html/avocado_result_html/__init__.py
@@ -198,7 +198,7 @@ class HTMLResult(Result):
 
     @staticmethod
     def _copy_static_resources(html_path):
-        module = 'avocado.plugins.html.avocado_result_html'
+        module = 'avocado_result_html'
         base_path = 'resources/static'
 
         for top_dir in pkg_resources.resource_listdir(module, base_path):
@@ -233,7 +233,7 @@ class HTMLResult(Result):
     def _render(self, result, output_path):
         context = ReportModel(result=result, html_output=output_path)
         template = pkg_resources.resource_string(
-            'avocado.plugins.html.avocado_result_html',
+            'avocado_result_html',
             'resources/templates/report.mustache')
 
         # pylint: disable=E0611

```

then I run `make link` and then `cd html; python setup.py develop --user`. Then it seems to work.

Comment (clebergnu) [inline html/MANIFEST.in]: Yep, we found it at the same time.  Pusing a hotfix (to be rebased).

File: avocado/core/html.py [hunk 2]
```diff
@@ -298,7 +298,7 @@ def _render_report(self):
                         template)
             ui.critical("-" * 80)
             ui.critical("%s:\n\n", details)
-            ui.critical("%r\n\n", self.json)
+            ui.critical("%r\n\n", self.result_dict)
```
Comment (ldoktor) [inline avocado/core/html.py]: Is this patch really necessary as you're removing the `result_dict` in the next commit?

File: avocado/core/job.py [hunk 3]
```diff
@@ -264,6 +264,8 @@ def _set_output_plugins(self):
             for klass in self.args.test_result_classes:
                 test_result_instance = klass(self)
                 self.result_proxy.add_output_plugin(test_result_instance)
+        else:
+            self.result_proxy.add_output_plugin(result.Result(self))
```
Comment (ldoktor) [inline avocado/core/job.py]: OK, makes sense, for now...

File: avocado/core/html.py [hunk 4]
```diff
@@ -1,289 +0,0 @@
...
-            if getattr(self.args, 'open_browser', False):
-                # if possible, put browser in separate process group, so
-                # keyboard interrupts don't affect browser as well as Python
-                setsid = getattr(os, 'setsid', None)
-                if not setsid:
-                    setsid = getattr(os, 'setpgrp', None)
-                inout = file(os.devnull, "r+")
-                cmd = ['xdg-open', self.output]
-                subprocess.Popen(cmd, close_fds=True, stdin=inout, stdout=inout,
-                                 stderr=inout, preexec_fn=setsid)
```
Comment (ldoktor) [inline avocado/core/html.py]: This is welcome. I'm only curious why it shows like this, did you use `git mv`?

Comment (ldoktor) [inline avocado/core/html.py]: Oh, you didn't, because it merges into the existing file... ignore my comment...

File: avocado/core/html.py (hunk 2 above)
Comment (clebergnu) [inline avocado/core/html.py]: IMHO it just makes a point: there was no JSON, it actually fixes that. And, it helps to understand the whole transition.

File: avocado/plugins/html.py [hunk 5]
```diff
@@ -15,13 +15,277 @@
...
+        open_browser = getattr(job.args, 'open_browser', False)
+        if getattr(job.args, 'html_job_result', 'off') == 'on':
+            html_dir = os.path.join(job.logdir, 'html')
+            os.makedirs(html_dir)
+            html_path = os.path.join(html_dir, 'results.html')
+            self._render(result, html_path)
+            if getattr(job.args, 'stdout_claimed_by', None) is None:
+                log = logging.getLogger("avocado.app")
+                log.info("JOB HTML   : %s", html_path)
+            if open_browser:
```
Comment (ldoktor) [inline avocado/plugins/html.py]: nice, this fixes the issue with multiple open browsers...

Comment (ldoktor) [inline avocado/plugins/html.py]: ... you should include the trello card in the changelog...

**Comment by clebergnu** [comment]
Closing in favor of #1375

File: html/MANIFEST.in (hunk 1 above)
Comment (ldoktor) [inline html/MANIFEST.in]: One more thing, I don't think we should put those real plugins directly into the main directory. How about creating `optional_plugins` directory and use `Makefile` to walk it and deploy them based on a variable?

```
make OPTIONAL_PLUGINS=html deploy
```
would install the html

Comment (clebergnu) [inline html/MANIFEST.in]: I'd propose something simpler at this point:

```diff
diff --git a/Makefile b/Makefile
index fcd15c3..fef6d9b 100644
--- a/Makefile
+++ b/Makefile
@@ -168,6 +168,7 @@ modules_boundaries:
 
 develop:
 	$(PYTHON) setup.py develop $(shell $(PYTHON26) || echo --user)
+   $(PYTHON) html/setup.py develop $(shell $(PYTHON26) || echo --user)
 
 link: develop
 	for MAKEFILE in $(AVOCADO_PLUGINS);\

```

Nothing prevents us from expanding on your idea, but I think we don't need anything more complex than this at this point.

PS: I honestly think we're using too much of `Makefile` in inspite of support already in `setup.py`, setuptools, pip, etc.  I'm going go send a few proposals to get more inline with the more standard Pythonic way of doing some of those things.

Comment (ldoktor) [inline html/MANIFEST.in]: Re to Makefile, yes, we might clean it up a bit, but in general Makefile is a standardized way to deploy/treat/tackle/test/... projects. It's independent to python and easy to use. So I would like to keep it as a primary way to `deploy`, `install`, create `rpms` and to run `selfchecks`.

File: avocado/core/html.py (hunk 2 above)
Comment (ldoktor) [inline avocado/core/html.py]: Sure, I don't mind that, it just looks unnecessary...

**Pull request closed without merging**